### ISSUE TITLE
fail2ban-client should automatically recover from 'database disk image is malformed'

### ISSUE THREAD

**Issue by dmak** [fail2ban/fail2ban#1465]
Fail2ban client logs the following into syslog:

> fail2ban-client[974]: ERROR  NOK: ('database disk image is malformed')

In general solution is provided [here](https://kb.plesk.com/en/127242), but if resetting the sqlite database is the only way to recover, I suppose that fail2ban should do it automatically (or try to recover the database as described [here](http://stackoverflow.com/questions/18259692/how-to-recover-a-corrupt-sqlite3-database)).

**Comment by sebres** [comment]
Thx for this.
I think, we can indeed try to restore or as failback just to recreate a database, if it went broken (after simple storing of backup).

Sounds like an easy implementing of FR. I'll later come to that (pressed for time)

**Comment by sebres** [comment]
Oh yes, forgotten, it would be especially important after merging of #1436, see https://github.com/fail2ban/fail2ban/pull/1436#issuecomment-220573533

**Comment by dmak** [comment]
Could you please tell if `fail2ban.sqlite3` is used to restore iptables bans during Fail2ban startup? If not (e.g. it is only used for "future" bans), then perhaps it makes sense to move it from `/var/lib/fail2ban` to `/run/fail2ban` (so that it is created each time the system is started)?

**Comment by sebres** [comment]
yes it is:
- used to restore bans
- used to find log file offset, since last search position after f2b restart
- you can use it for other purposes, like get all failures of any IP any jail... for example to generate abuse

but if you don't want or need to have it persistent - you can place it there as suggested

**Comment by dmak** [comment]
Thanks for quick reply. It is placed where it should be. I have added this file also to backup cycle, so I can restore it / fallback in case of a problem like being discussed.

**Comment by RobbieTheK** [comment]
Not sure if this is a regression but Fedora 33, `fail2ban-0.11.2-1.fc33.noarch`

`Caught unhandled exception in main cycle: DatabaseError('database disk image is malformed`

Do I just move  `/var/lib/fail2ban/fail2ban.sqlite3 and restart F2B`?

**Comment by sebres** [comment]
When is it happened exactly, during the run-time? 

#2004 (included in 0.11.2) allows automatically repair by start of fail2ban, so you don't need to move it if this error will be recognized by connect to DB after restart.
Otherwise (if it is not help and you don't see `Trying to repair database` in fail2ban-log) - the answer is yes.

**Comment by RobbieTheK** [comment]
> When is it happened exactly, during the run-time?

The first time it occurred it was after a restart of F2B and during the re-banning process.

```
2021-03-14 00:45:09,614 fail2ban.actions        [1000]: NOTICE  [sshd] Unban 104.131.190.193
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
2021-03-14 00:47:20,671 fail2ban.server         [1051]: INFO    --------------------------------------------------
2021-03-14 00:47:20,698 fail2ban.server         [1051]: INFO    Starting Fail2ban v0.11.2
2021-03-14 00:47:20,698 fail2ban.observer       [1051]: INFO    Observer start...
2021-03-14 00:47:21,016 fail2ban.database       [1051]: INFO    Connected to fail2ban persistent database '/var/lib/fail2ban/fail2ban.sqlite3'
2021-03-14 00:47:21,035 fail2ban.jail           [1051]: INFO    Creating new jail 'sshd'
2021-03-14 00:47:43,898 fail2ban.actions        [1051]: NOTICE  [recidive] Restore Ban 103.109.152.210
2021-03-14 00:47:43,986 fail2ban.filter         [1051]: WARNING [sshd] Simulate NOW in operation since found time has too large deviation 1615700761.384607 ~ 1615700863.986449 +/- 60
2021-03-14 00:47:43,986 fail2ban.filter         [1051]: WARNING [sshd] Please check jail has possibly a timezone issue. Line with odd timestamp: ('', '2021-03-14T00:46:01.384607', 'ourdomain.edu sshd[999]: Received signal 15; terminating.')
2021-03-14 00:47:43,988 fail2ban.filter         [1051]: INFO    [sshd] Found 104.131.190.193 - 2021-03-14 00:47:30
2021-03-14 00:47:43,989 fail2ban.filter         [1051]: INFO    [sshd] Found 104.131.190.193 - 2021-03-14 00:47:31
2021-03-14 00:47:43,989 fail2ban.observer       [1051]: INFO    [sshd] Found 104.131.190.193, bad - 2021-03-14 00:47:30, 1 # -> 2.0
2021-03-14 00:47:44,025 fail2ban.filtersystemd  [1051]: ERROR   Caught unhandled exception in main cycle: DatabaseError('database disk image is malformed')
2021-03-14 00:47:44,026 fail2ban.observer       [1051]: INFO    [sshd] Found 104.131.190.193, bad - 2021-03-14 00:47:31, 1 # -> 2.0
2021-03-14 00:47:44,137 fail2ban.actions        [1051]: NOTICE  [recidive] Restore Ban 103.130.213.21
2021-03-14 00:47:44,381 fail2ban.actions        [1051]: NOTICE  [recidive] Restore Ban 103.131.169.149
2021-03-14 00:47:44,629 fail2ban.actions        [1051]: NOTICE  [recidive] Restore Ban 103.136.40.90
2021-03-14 00:47:44,868 fail2ban.actions        [1051]: NOTICE  [recidive] Restore Ban 103.145.22.49
2021-03-14 00:47:45,133 fail2ban.actions        [1051]: NOTICE  [recidive] Restore Ban 103.213.96.15
2021-03-14 00:47:45,499 fail2ban.actions        [1051]: NOTICE  [recidive] Restore Ban 103.214.129.204
2021-03-14 00:47:46,004 fail2ban.filter         [1051]: WARNING [pam-generic] Simulate NOW in operation since found time has too large deviation 1615700762.388586 ~ 1615700866.0043046 +/- 60
2021-03-14 00:47:46,004 fail2ban.filter         [1051]: WARNING [pam-generic] Please check jail has possibly a timezone issue. Line with odd timestamp: ('', '2021-03-14T00:46:02.388586', 'ourdomain.edu dovecot[1191]: master: Warning: Killed with signal 15 (by pid=103672 uid=0 code=kill)')
2021-03-14 00:47:46,161 fail2ban.actions        [1051]: NOTICE  [recidive] Restore Ban 103.219.147.34
```

 Perhaps all of those[ ^@](url) caused a corruption?



> Otherwise (if it is not help and you don't see `Trying to repair database` in fail2ban-log)

I do not see that in any of the logs, so I moved the old DB out of the way and F2B recreated it.
